sdks/python/apache_beam/runners/common.py

```diff
@@ -51,6 +51,270 @@ def receive(self, windowed_value):
     raise NotImplementedError
 
 
+class DoFnMethodWrapper(object):
+  """Represents a method of a DoFn object."""
+
+  def __init__(self, do_fn, method_name):
+    """
+    Initiates a ``DoFnMethodWrapper``.
+
+    Args:
+      do_fn: A DoFn object that contains the method.
+      method_name: name of the method as a string.
+    """
+
+    args, _, _, defaults = do_fn.get_function_arguments(method_name)
+    defaults = defaults if defaults else []
+    method_value = getattr(do_fn, method_name)
+    self.method_value = method_value
+    self.args = args
+    self.defaults = defaults
+
+
+class DoFnSignature(object):
+  """Represents the signature of a given ``DoFn`` object.
+
+  Signature of a ``DoFn`` provides a view of the properties of a given ``DoFn``.
+  Among other things, this will give an extensible way for for (1) accessing the
+  structure of the ``DoFn`` including methods and method parameters
+  (2) identifying features that a given ``DoFn`` support, for example, whether
+  a given ``DoFn`` is a Splittable ``DoFn`` (
+  https://s.apache.org/splittable-do-fn) (3) validating a ``DoFn`` based on the
+  feature set offered by it.
+  """
+
+  def __init__(self, do_fn):
+    # We add a property here for all methods defined by Beam DoFn features.
+
+    assert isinstance(do_fn, core.DoFn)
+    self.do_fn = do_fn
+
+    self.process_method = DoFnMethodWrapper(do_fn, 'process')
+    self.start_bundle_method = DoFnMethodWrapper(do_fn, 'start_bundle')
+    self.finish_bundle_method = DoFnMethodWrapper(do_fn, 'finish_bundle')
+    self._validate()
+
+  def _validate(self):
+    # start_bundle and finish_bundle methods should only have ContextParam as a
+    # default argument.
+    self._validate_start_bundle()
+    self._validate_finish_bundle()
+    self._validate_process()
+
+  def _validate_start_bundle(self):
+    self._validate_bundle_method(self.start_bundle_method)
+
+  def _validate_finish_bundle(self):
+    self._validate_bundle_method(self.finish_bundle_method)
+
+  def _validate_process(self):
+    pass
+
+  def _validate_bundle_method(self, method_wrapper):
+    # Bundle methods may only contain ContextParam.
+    unsupported_dofn_params = [i for i in core.DoFn.__dict__ if (
+        i.endswith('Param') and i != 'ContextParam')]
+
+    for param in unsupported_dofn_params:
+      assert param not in method_wrapper.defaults
```

Comment:
Here you're implicitly using the fact that XxxParam is both named XxxParam and has value "XxxParam." At least call this out.